File: 목요빈/Week15/PRGR_가장많이받은선물.java
import java.util.*;

class PRGR_가장많이받은선물 {
    
    /*
    선물을 주고받은 기록을 바탕으로 다음달에 누가 선물을 많이 받을지 예측
    
    - 두 사람이 선물을 주고 받은 기록이 있으면 더 많은 선물을 준 사람이 다음달에 선물을 더 받음
    - 하나도 없거나 같다면 선물 지수가 더 큰 사람이 받음, 선물지수도 같으면 다음 달에 주고받지 않음
    * 선물지수 = 이번 달까지 자신이 친구들에게 준 선물의 수 - 받은 수 == 많이 줬을수록 높음
    
    선물을 가장 많이 받을 친구는?
    
    그래프?
    사람별로 준사람, 받은사람 저장 -> 맵 이용해서 숫자로 저장
    
    */
    int N;
    
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        N = friends.length;
        Map<String, Integer> map = new HashMap<>();
        for(int i = 0; i<friends.length; i++){
            map.put(friends[i], i);
        }
        
        int[][] graph = new int[N][N];
        int[] degree = new int[N];
        for(int i = 0; i<gifts.length; i++){
            String[] gift = gifts[i].split(" ");
            degree[map.get(gift[0])] += 1;
            degree[map.get(gift[1])] -= 1;
            graph[map.get(gift[0])][map.get(gift[1])] += 1;
        }
        
        for(int i = 0; i<N; i++){
            int temp = 0;
            for(int j = 0; j<N; j++){
                if(i == j) continue;
                if(graph[i][j] > graph[j][i] || (graph[i][j] == graph[j][i] && degree[i] > degree[j])){
                    temp += 1;
                }
            }
            
            answer = Math.max(answer, temp);
        }
        return answer;
    }
}

File: 전영주/Week15/PRGR_가장많이받은선물.js
function solution(friends, gifts) {
  let answer = 0;
  let size = friends.length;
  let mapping = new Map();
  let gift = Array.from({ length: size }, e => Array.from({ length: size }, e => 0));
  let gValue = Array.from({ length: size }, e => Array.from({ length: 2 }, e => 0));//주고 받고
  friends.forEach((e, i) => {
    mapping.set(e, i);
  })
  //이번달 주고 받은거
  gifts.forEach(e => {
    let [a, b] = e.split(" ");//a to b
    let aI = mapping.get(a);
    let bI = mapping.get(b);
    gift[bI][aI]++;
    gValue[aI][0]++;
    gValue[bI][1]++;
  })
  //다음달 계산
  let nextGift = Array.from({ length: size }, e => 0);
  for (let i = 0; i < size; i++) {
    for (let j = i + 1; j < size; j++) {
      if (gift[i][j] > gift[j][i]) {//i가 j에게 받은 수가 더 크면
        nextGift[j]++;
      } else if (gift[i][j] < gift[j][i]) {//j가 i에게 받은 수가 더 크면
        nextGift[i]++;
      } else {
        if (gValue[i][0] - gValue[i][1] > gValue[j][0] - gValue[j][1]) {
          nextGift[i]++;
        } else if (gValue[i][0] - gValue[i][1] < gValue[j][0] - gValue[j][1]) {
          nextGift[j]++;
        }
      }
    }
  }
  return Math.max(...nextGift);
}
/*
1대 1 관계에서
-더많이 준 사람이 다음달 선물 하나 받음
-선물지수(준거-받은거)가 더 큰 사람이 작은 사람에게 하나 받음 
선물 지수도 같으면 다음달 ㄴㄴ
가장 많이 받을 친구가 받은 선물 수?

1. 받은 수: arr[friends x][friends y] x가 y에게 받은 수
2. 준 수: arr[friends y][friends x] x가 y에게 준 수
한 사람이 받은 수는 arr[x]의 원소 합
50이라 가능

24분 1코에 성공
*/